### PR TITLE
Adds Http, https, and ftp support to protect (resolves #149)

### LINKED ISSUE #149: Add support for https download links
Pipelines should be able to download files form https links (Redwood will probably give us signed https links).

### PR DESCRIPTION
**Pull request by akul-goyal**

### CODE REVIEW

**Comment by arkal** [comment]
Failed due to Transgene docker mislabelling. Fixing in #158.

**Comment by arkal** [comment]
Hey @akul-goyal if you rebase to the latest master the ci test will pass.

**Comment by akul-goyal** [comment]
Hi Arjun,

I rebased the changes onto master.

Let me know if you have any problems.

Akul

**Comment by arkal** [comment]
Hi @akul-goyal , I don't think your rebase worked, the test failed since it is still looking for Transgene:1.0.0 and not 2.0.0. We'll look at it tomorrow.

**Comment by akul-goyal** [comment]
jenkins, test this please

**Comment by arkal** [comment]
Hey @akul have you changed the policy on the bucket yet? This will fail as long as the policy isn't set to the new one.

**Comment by akul-goyal** [comment]
Hi Arjun,

I did change the policy and in the test which I ran on one of the files in
the bucket, it worked.

Akul

**Comment by arkal** [comment]
This was failing since the bucket was set up as requester pays and anonymous downloads are not permitted from requester-pays buckets. @akul-goyal will add a fix to catch the error that allowed the previous runs to move forward after the failed download and resubmit this PR WITHOUT the https input in the CI test inputs. I will re-run tests with the unit-test make target for jenkins and merge it in pending those tests passing.

File: src/protect/pipeline/ProTECT.py [hunk 1]
```diff
@@ -500,11 +502,9 @@ def get_pipeline_inputs(job, input_flag, input_file, encryption_key=None,
...
                               os.path.basename(input_file))
-    if input_file.startswith('http'):
-        assert input_file.startswith('https://s3'), input_file + ' is not an S3 file'
-        input_file = get_file_from_s3(job, input_file, write_to_jobstore=False,
-                                      encryption_key=encryption_key,
-                                      per_file_encryption=per_file_encryption)
+    if input_file.startswith(('http', 'https', 'ftp')):
+        input_file=get_file_from_url(job, input_file,encryption_key=encryption_key,
+                        per_file_encryption=per_file_encryption, write_to_jobstore=False)
     elif input_file.startswith('S3'):
```
Comment (arkal) [inline src/protect/pipeline/ProTECT.py]: On the next line  ` elif input_file.startswith(('s3', 'S3')):`

File: src/protect/test/unit/test_file_downloads.py [hunk 2]
```diff
@@ -102,5 +103,49 @@ def _download_files(job):
...
+    @staticmethod
+    def _download_https_files(job):
+        """
+        Attempts to download an unencrypted file, a file encrypted with a key, and a file encrypted
+        with a hash of a master key.
+        """
+        http_base ='https://s3-us-west-2.amazonaws.com/cgl-protect-data/ci_references/rsem_index_chr6.tar.gz'
+        try:
+            get_file_from_url(job, http_base, write_to_jobstore=False)
+        except RuntimeError as err:
```
Comment (arkal) [inline src/protect/test/unit/test_file_downloads.py]: You are handling this is get_file_from url, right?

File: src/protect/test/unit/test_file_downloads.py [hunk 3]
```diff
@@ -102,5 +103,49 @@ def _download_files(job):
...
+    def test_file_downloads_from_https(self):
+        """
+        Test the functionality of get_file_from_url https
+        """
+        a = Job.wrapJobFn(self._download_https_files)
+        Job.Runner.startToil(a, self.options)
+
+
+    @staticmethod
+    def _download_https_files(job):
```
Comment (arkal) [inline src/protect/test/unit/test_file_downloads.py]: This test is currently downloads the https s3 link using get_file from s3. Rename this test to be more informational and then write an additional test that tests an https link that is downloaded directly with urllib.urlopen2

File: src/protect/test/unit/test_file_downloads.py [hunk 4]
```diff
@@ -102,5 +103,49 @@ def _download_files(job):
...
+    @staticmethod
+    def _download_http_files(job):
+        """
+        Attempts to download an unencrypted file, a file encrypted with a key, and a file encrypted
+        with a hash of a master key.
+        """
+        http_base = 'http://link.aps.org'
+        try:
+            get_file_from_url(job, http_base, write_to_jobstore=False)
+        except RuntimeError as err:
```
Comment (arkal) [inline src/protect/test/unit/test_file_downloads.py]: see above

File: src/protect/test/unit/test_file_downloads.py (hunk 3 above)
Comment (akul-goyal) [inline src/protect/test/unit/test_file_downloads.py]: Do you want me to also test for an error ie: put a dummy link?

**Comment by arkal** [comment]
Yes!

**Comment by arkal** [comment]
Jenkins, test this please.

File: src/protect/common.py [hunk 5]
```diff
@@ -190,7 +192,7 @@ def get_file_from_s3(job, s3_url, encryption_key=None, per_file_encryption=True,
     parsed_url = urlparse(s3_url)
     if parsed_url.scheme == 'https':
         download_url = 'S3:/' + parsed_url.path  # path contains the second /
-    elif parsed_url.scheme == 's3':
+    elif parsed_url.scheme == 's3' or parsed_url.scheme == 'S3':
```
Comment (arkal) [inline src/protect/common.py]: `parsed_url.scheme in ('s3', 'S3')`

File: src/protect/common.py [hunk 6]
```diff
@@ -256,6 +258,42 @@ def get_file_from_s3(job, s3_url, encryption_key=None, per_file_encryption=True,
...
+    work_dir = job.fileStore.getLocalTempDir()
+
+    filename = '/'.join([work_dir, str(uuid.uuid4())])
+    url = any_url
+    parsed_url = urlparse(any_url)
+    try:
+        response = urllib2.urlopen(url)
+
+    except urllib2.HTTPError:
+        if (parsed_url.netloc).startswith('s3') or (parsed_url.netloc).startswith('S3'):
```
Comment (arkal) [inline src/protect/common.py]: `parsed_url.netloc.startswith(('s3', 'S3'))`

File: src/protect/common.py [hunk 7]
```diff
@@ -256,6 +258,42 @@ def get_file_from_s3(job, s3_url, encryption_key=None, per_file_encryption=True,
...
+            job.fileStore.logToMaster("Detected https link is for a encrypted s3 file.")
```
Comment (arkal) [inline src/protect/common.py]: an encrypted

File: src/protect/common.py [hunk 8]
```diff
@@ -256,6 +258,42 @@ def get_file_from_s3(job, s3_url, encryption_key=None, per_file_encryption=True,
...
+            filename = get_file_from_s3(job, any_url, encryption_key=encryption_key,
+                                        per_file_encryption=per_file_encryption,
+                                        write_to_jobstore=False)
```
Comment (arkal) [inline src/protect/common.py]: `write_to_jobstore=write_to_jobstore`. get_file_from _s3 should get the same value sent to get_file_from_url

File: src/protect/pipeline/ProTECT.py (hunk 1 above)
Comment (arkal) [inline src/protect/pipeline/ProTECT.py]: This has not been addressed

File: src/protect/test/unit/test_file_downloads.py [hunk 9]
```diff
@@ -101,6 +102,94 @@ def _download_files(job):
...
+
+    def test_file_downloads_from_https_fail(self):
+        """
+        Test the functionality of get_file_from_url https
+        """
+        a = Job.wrapJobFn(self._download_https_files_fail)
+        Job.Runner.startToil(a, self.options)
+
+    @staticmethod
+    def _download_https_files_fail(job):
```
Comment (arkal) [inline src/protect/test/unit/test_file_downloads.py]: on hindsight, drop this test as it is basically testing urllib2 functionality

File: src/protect/common.py (hunk 5 above)
Comment (akul-goyal) [inline src/protect/common.py]: test fails is I add S3

File: src/protect/common.py [hunk 10]
```diff
@@ -256,6 +258,42 @@ def get_file_from_s3(job, s3_url, encryption_key=None, per_file_encryption=True,
...
+    filename = '/'.join([work_dir, str(uuid.uuid4())])
+    url = any_url
+    parsed_url = urlparse(any_url)
+    try:
+        response = urllib2.urlopen(url)
+
+    except urllib2.HTTPError:
+        if (parsed_url.netloc).startswith(('s3', 'S3')):
+            job.fileStore.logToMaster("Detected https link is for an encrypted s3 file.")
+            filename = get_file_from_s3(job, any_url, encryption_key=encryption_key,
```
Comment (arkal) [inline src/protect/common.py]: `return get_file_from_s3(...)`

**Pull request merged**